### ISSUE TITLE
Random words seem work too.

### ISSUE THREAD

**Issue by kai-wen-yang** [umd-huang-lab/perceptionCLIP#1]
Hi, I build a factor template using random words and add it to "imagenet_factor_templates":

```
    "helloworld": [
        "",
        ", hello",
        ", world",
        ", !",
    ],

```

Then I use this factor template to run zero-shot generalization:

```
python src/zero_shot_inference/perceptionclip_two_step.py --model=ViT-B/16 --dataset=ImageNet --main_template=imagenet_main_template --factor_templates=imagenet_factor_templates --factors=helloworld --infer_mode=0 --temperature=3 --save_name=imagnet_ours_wy_vit16
```

It achieves 68.1% accuracy, outperforming most of the factors like background and orientation. 
How to explain this phenomenon? Does "hello world!" provide context information? Or does the accuracy gain all come from ensemble with random words?

**Comment by bangann** [comment]
Hi Kaiwen,

This is a great example to discuss. Thank you for bringing it up! Here're our understandings of this phenomenon.

TLDR: {"", ", hello", ", world", ", !"} is less effective on other datasets, so we conjecture that its performance gain on ImageNet mainly stems from the text augmentation effect (as demonstrated in [Roth et al.](https://arxiv.org/abs/2306.07282) and our Table 2), and the diverse images in ImageNet could potentially confound the effects of text augmentation and the contextual attributes.

1) We observed that using {"", ", hello", ", world", ", !"} does not always help. It is less effective on other datasets, especially when compared to using contextual attributes.
    The Waterbirds dataset has a clear (spurious) attribute — background. The group robustness of "helloworld" on this dataset shows little improvement compared to the original CLIP and is significantly lower than the performance of using the "background" attribute.
    | waterbirds          |   RN50  |           |           | ViT-B/32 |           |           | ViT-B/16 |           |           | ViT-B/14 |           |           |
    |---------------------|:-------:|:---------:|:---------:|:--------:|:---------:|:---------:|:--------:|:---------:|:---------:|:--------:|:---------:|:---------:|
    |                     | avg     | worst     | gap       | avg      | worst     | gap       | avg      | worst     | gap       | avg      | worst     | gap       |
    | CLIP                | 90.47   | 16.07     | 74.4      | 87.34    | 47.28     | 40.06     | 87.34    | 26.79     | 60.56     | 90.55    | 44.64     | 45.91     |
    | **+background** | 90.01   | **32.14** | **57.87** | 78.6     | **60.33** | **18.28** | 85.8     | **41.07** | **44.73** | 87.74    | **61.12** | **26.62** |
    | **+helloworld**  |  89.82  |  16.07    |  73.75    |  87.67   |  45.65    |  42.02    |  87.90   |  30.36    |  57.55    |  89.89   |  46.43    |  43.46    |

    On other datasets, the performance gain achieved using "helloworld" does not exceed half of the performance gain brought by using contextual attributes, except for "Places365". Notably, it hurts the performance on EuroSAT.
    |                 | CUB       | EuroSAT| Places365 | Flowers102 | Food101   | Oxford Pets |
    |-----------------|-----------|-----------|-----------|------------|-----------|------------|
    | domain_template | 56.32     | 54.94     | 38.93     | 70.99      | 88.72     | 89.04      |
    | **+attributes** | **57.08** | **59.23** | **40.27** | **72.84**  | **89.14** | **90.37**  |
    | **+helloworld** |  56.37    |  54.23    |  39.96    |  71.69     |  88.94    |  89.72     |

    On all out-of-distribution versions of ImageNet, "helloworld" performs worse than "background" but better than "orientation." Note that our "orientation" attribute only consists of three values (descriptions): "", "upside-down," and "rotated." This attribute hardly provides any information since ImageNet images are mostly upright. We include this attribute in Table 4 mainly for the purpose of demonstrating the effects of different attributes.
    |                     | ImageNet | ImageNetV2 | ImageNet-R | ImageNet-A | ImageNet-Sketch |
    |---------------------|----------|------------|------------|------------|-----------------|
    | simple template               | 66.72    | 60.85      | 73.99      | 47.8       | 46.16           |
    | **+helloworld**     | 68.16    | 61.63      | 74.95      | 48.92      | 46.74           |
    | **+background**     | 67.70     | 61.91      | 75.71      | 49.13      | 47.21           |
    | **+orientation**    | 67.21    | 61.04      | 74.31      | 47.88      | 46.54           |
    | **+best attribute** | 68.30     | 61.91      | 77.08      | 50.41      | 47.95           |
2) Text augmentation can indeed improve CLIP's inference performance, which could be the main reason of why "helloworld" is effective. As shown in Table 2 of our paper, the performance gain achieved by conditioning on randomly generated words is approximately half of that obtained by using ground-truth attributes. The effectiveness of text augmentation may be due to CLIP's sensitivity to text perturbations.
3) Another possible reason for the effectiveness of "helloworld" might be the high diversity of ImageNet's images. It's hard to say whether the text descriptions like "hello", "world" and "!" are not part of some descriptions of some generative attributes. This is why, in Section 5, we manually adjusted the images and their corresponding attributes to reduce such confounding influences.

We hope this can, to some extent, explain this intriguing observation. Happy to discuss more!

**Comment by kai-wen-yang** [comment]
Thanks for the reply.
